storage/repository/generic.go

```diff
@@ -0,0 +1,128 @@
+package repository
+
+import (
+	"fmt"
+	"strings"
+
+	"fyne.io/fyne"
+)
+
+// GenericParent can be used as a common-case implementation of
+// HierarchicalRepository.Parent(). It will create a parent URI based on
+// IETF RFC3986.
+//
+// In short, the URI is separated into it's component parts, the path component
+// is split along instances of '/', and the trailing element is removed. The
+// result is concatenated and parsed as a new URI.
+//
+// If the URI path is empty or '/', then a duplicate of the URI is returned,
+// along with URIRootError.
+//
+// The parseURI parameter should be a function that you wish GenericParent to
+// use to generate any new URIs it creates. Usually, storage.ParseURI is
+// suitable here.
+//
+// NOTE: this function should not be called except by an implementation of
+// the Repository interface - using this for unknown URIs may break.
+//
+// Since 2.0.0
+func GenericParent(u fyne.URI, parseURI func(string) (fyne.URI, error)) (fyne.URI, error) {
+
+	if u.Path() == "" || u.Path() == "/" {
+		parent, err := parseURI(u.String())
+		if err != nil {
+			return nil, err
+		}
+		return parent, URIRootError
+	}
+
+	components := strings.Split(u.Path(), "/")
+
+	newURI := u.Scheme() + "://" + u.Authority()
+
+	// there will be at least one component, since we know we don't have
+	// '/' or ''.
+	if len(components) == 1 {
+		// the immediate parent is the root
+		newURI += "/"
+	} else {
+		newURI += strings.Join(components[:len(components)-1], "/")
+	}
+
+	// stick the query and fragment back on the end
+	if len(u.Query()) > 0 {
+		newURI += "?" + u.Query()
```

Comment:
nit: assign to variable, ditto for fragment and in GenericChild below
```suggestion
	if q := u.Query(); len(q) > 0 {
		newURI += "?" + q
```

Reply:
26e7a0e31c5018fff46f5d15980ee52527291fd2